sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypes.scala

```diff
@@ -142,3 +142,29 @@ case class CreateArray(children: Seq[Expression]) extends Expression {
 
   override def toString: String = s"Array(${children.mkString(",")})"
 }
+
+/**
+ * Returns a Row containing the evaluation of all children expressions.
+ */
+case class CreateStruct(children: Seq[Expression]) extends Expression {
+  override type EvaluatedType = Row
+
+  override def foldable: Boolean = children.forall(_.foldable)
+
+  override lazy val resolved: Boolean = childrenResolved
+
+  override def dataType: StructType = {
+    assert(resolved, s"CreateStruct is called with unresolved children: $children.")
```

Comment:
The error message is confusing. `CreateStruct` *can* be called with unresolved children (and you actually did this [here] [1]), but `CreateStruct.dataType` can't, because child types may be unknown.

[1]: https://github.com/apache/spark/pull/5196/files#diff-9c84e1d27f25714e256cb482069359cfR52